common/events/eventbus.go

```diff
@@ -0,0 +1,171 @@
+// Package events is the event center for internal system events.
+// All modules could register event callback for each event.
+package events
+
+import (
+	"errors"
+	"reflect"
+	"sync"
+)
+
+// Event center errors
+var (
+	ErrInvalidParams    = errors.New("Invalid Params")
+	ErrQueueFull        = errors.New("Event queue is full")
+	ErrDuplicateHandler = errors.New("Duplicate handler")
+	ErrInvalidHandler   = errors.New("Invalid handler")
+	ErrPartialSuccess   = errors.New("Operations partial success")
+)
+
+// EventBus is the event hub for all registered events
+type EventBus struct {
+	handler map[EventType][]EventHandler
+	lock    sync.RWMutex
+}
+
+var eventBus *EventBus
+
+func newEventBus() *EventBus {
+	return &EventBus{
+		make(map[EventType][]EventHandler),
+		sync.RWMutex{},
+	}
+}
+
+func init() {
+	eventBus = newEventBus()
+}
+
+// GetEventBus return the instance of EventBus
+func GetEventBus() *EventBus {
+	return eventBus
+}
+
+func (ec *EventBus) findHandlerIndex(eType EventType, handler EventHandler) int {
+	if ehs, ok := ec.handler[eType]; ok {
+		shValue := reflect.ValueOf(handler)
+		for idx, eh := range ehs {
+			ehValue := reflect.ValueOf(eh)
+			if shValue.Pointer() == ehValue.Pointer() {
+				return idx
+			}
+		}
+	}
+	return -1
+}
+
+// Subscribe specified event type with a handler
+func (ec *EventBus) Subscribe(eType EventType, handler EventHandler) error {
+	if handler == nil {
+		return ErrInvalidParams
+	}
+	ec.lock.Lock()
+	defer ec.lock.Unlock()
+	if ec.findHandlerIndex(eType, handler) >= 0 {
+		// handler already exists
+		return ErrDuplicateHandler
+	}
+
+	if _, ok := ec.handler[eType]; !ok {
+		ec.handler[eType] = make([]EventHandler, 0)
+	}
+	ec.handler[eType] = append(ec.handler[eType], handler)
+	return nil
+}
+
+// SubscribeMulti subscribe multiple events with one handler
+func (ec *EventBus) SubscribeMulti(eTypes []EventType, handler EventHandler) ([]EventType, error) {
+	if len(eTypes) == 0 || handler == nil {
+		return nil, ErrInvalidParams
+	}
+	failedTypes := []EventType{}
+	for _, eType := range eTypes {
+		err := ec.Subscribe(eType, handler)
+		if err != nil {
+			failedTypes = append(failedTypes, eType)
+		}
+	}
+	if len(failedTypes) != 0 {
+		return failedTypes, ErrPartialSuccess
+	}
+	return nil, nil
+}
+
+// Unsubscribe the given event handler of specified event type
+func (ec *EventBus) Unsubscribe(eType EventType, handler EventHandler) error {
+	if handler == nil {
+		return ErrInvalidParams
+	}
+	ec.lock.Lock()
+	defer ec.lock.Unlock()
+	idx := ec.findHandlerIndex(eType, handler)
+	if idx < 0 {
+		// handler not found, treat as unsubscribe successfully
+		return ErrInvalidHandler
```

Comment:
Add HandlerNotExist = errors.New("handler not exist"),  this error description is more appropriate for the scene

Reply:
I think the return value should be nil in this case, just like the comments "handler not found, treat as unsubscribe successfully".